controllers/indexer.py

```diff
@@ -60,35 +62,49 @@ def intaking(self) -> None:
     def reading(self, state_tm: float) -> None:
         self.indexer.read_cargo_colour()
         colour = self.indexer.get_cargo_colour()
-        if state_tm > 0.3:
-            if not colour.is_valid() and state_tm < 0.5:
-                return
-            self.log_colour.append(colour.name)
-            if colour.is_opposition() and not self.ignore_colour:
-                if self.catflap_active:
-                    if (
-                        self.indexer.has_trapped_cargo
-                        or self.indexer.has_cargo_in_chimney()
-                    ):
-                        self.next_state("clearing")
-                    else:
-                        self.next_state("trapping")
-                else:
-                    self.next_state("clearing")
+        if state_tm < 0.3 or (not colour.is_valid() and state_tm < 0.5):
+            return
+
+        self.log_colour.append(colour.name)
+        if colour.is_opposition() and not self.ignore_colour:
+            if (
+                self.catflap_active
+                and not self.indexer.has_trapped_cargo
+                and not self.indexer.has_cargo_in_chimney()
+            ):
+                self.next_state("trapping")
             else:
-                if self.catflap_active and self.ignore_colour:
-                    if self.indexer.has_trapped_cargo:
-                        self.next_state("clearing")
-                    else:
-                        self.next_state("trapping")
+                self.next_state("clearing")
+        else:  # our ball or ignore colour
+            if self.catflap_active and self.ignore_colour:
+                if self.indexer.has_trapped_cargo:
+                    self.next_state("clearing")
                 else:
-                    # It is our ball so we have finished this process
-                    # The "stopped" state will work out if it needs to move the ball into the chimney
-                    self.next_state("stopped")
+                    self.next_state("trapping")
+            else:
+                # It is our ball so we have finished this process
+                # The "stopped" state will work out if it needs to move the ball into the chimney
+                self.next_state("stopped")
+
+    @state(must_finish=True)
+    def clearing(self, initial_call) -> None:
+        if initial_call:
+            if self.indexer.has_cargo_in_chimney():
+                self.next_state("intake_clearing")
+            else:
+                self.shooter_control.clear()
+
+        if not (
+            self.shooter_control.current_state == "prepare_to_clear"
+            or self.shooter_control.current_state == "clearing"
+        ):
+            self.next_state("stopped")
```

Comment:
Will this cause it to never clear through the shooter?
 * Initial call happens
 * Shooter controller set to clear -> this just sets a private flag in the controller
 * Lines 97-100 check the state of the shooter controller, but it hasn't changed yet, so the indexer moves to the `stopped` state